js-pkg/demos/webrtc-demo/webrtc.js

```diff
@@ -0,0 +1,181 @@
+import * as React from 'react'
+import { useDriftDBSignalingChannel } from './driftdbutils';
+
+
+export const useWebRTCMessagingChannel = (p1, p2) => {
+    let [isInited, isReady, connRef, connID] = useWebRTCConnection(p1,p2)
+    let [theyAreReady, setTheyAreReady] = React.useState(false)
+    let [messages, addMessage] = React.useReducer((state, msg) => [...state, msg], [])
+    let mydataChannelRef = React.useRef(null)
+    let theirdataChannelRef = React.useRef(null)
+    React.useEffect(() => {
+	if(isInited) {
+	    let dataChannel = connRef.current.createDataChannel(p1)
+	    dataChannel.onopen = () => {
+		console.log("DC OPEN!!!")
+	    }
+
+	    dataChannel.onmessage = e => {
+		addMessage({id: e.timeStamp, text: e.data})
+	    }
+
+
+	    dataChannel.onerror = e => {
+		console.error(e)
+	    }
+
+
+	    connRef.current.ondatachannel = (e) => {
+		console.log("ondc fires")
+		if(e.channel.label === p2) {
+		    setTheyAreReady(true)
+		    theirdataChannelRef.current = e.channel
+		}
+	    }
+		    
+	    mydataChannelRef.current = dataChannel
+	}
+	return () => {
+	    setTheyAreReady(false)
+	    if(isReady) {
+		mydataChannelRef.current.close()
+	    }
+	}
+    }, [connID, isInited])
+
+    return [ messages, theyAreReady ? theirdataChannelRef.current.send.bind(theirdataChannelRef.current) : (msg) => {console.log("unsent: ", msg)} ]
+}
+
+export const useWebRTCConnection = (p1, p2) => {
+    let [signalingMessages, setSignalingMessages, removeFromRecv, emptySend]= useDriftDBSignalingChannel(p1, p2)
+    const [connID, setconnID] = React.useState(0)
+    const incrementConnID = ()=>setconnID((connID) => connID + 1)
+    const [isReady, setIsReady] = React.useState(false)
+    const reachedStableRef = React.useRef(false)
+    const [isInited, setIsInited] = React.useState(false)
+    const isPeerPolite = p1 < p2;
+    let connRef = React.useRef(null)
+    let makingOfferRef = React.useRef(false)
+    React.useEffect(() => {
+	let conn = new RTCPeerConnection({
+	    iceServers: [
+		{urls: "stun:stun.l.google.com:19302"}
+	    ]}) 
+
+	connRef.current = conn
+	setIsInited(true)
+
+	conn.addEventListener("negotiationneeded", async () => { 
+		    try {
+			makingOfferRef.current = true
+			await conn.setLocalDescription()
+			setSignalingMessages({ type: "offer", sdp: conn.localDescription.sdp })
+		    } catch(e) { console.error(e) } finally {
+			makingOfferRef.current = false 
+		    }
+	})
+
+	conn.onicecandidate = (e) => {
```

Comment:
Curious about the mixed usage of `addEventListener('eventname')` and `oneventname = () => { ... }`. Is that intentional?

Reply:
nope, I had two listeners on that event at one point and so changed it, will change it back.